lib/ansible/modules/remote_management/manageiq/manageiq_alert_profiles.py

```diff
@@ -0,0 +1,348 @@
+        except Exception as e:
+            self.module.fail_json(msg="Deleting profile failed: {error}".format(error=e))
+
+        msg = "Successfuly deleted profile {description}"
+        msg = msg.format(description=profile['description'])
+        return dict(changed=True, msg=msg)
+
+    def get_alert_policy_href(self, alert_policy):
+        """ Get an absolute href for an alert policy
+        """
+        return "{url}/alert_definitions/{id}".format(url=self.api_url, id=alert_policy['id'])
+
+    def assign_or_unassign(self, profile, resources, action):
+        """ Assign or unassign alerts to profile, and validate the result.
+        """
+        assert action in ["assign", "unassign"]
+
+        subcollection_url = profile['href'] + '/alert_definitions'
+        try:
+            result = self.client.post(subcollection_url, resources=resources, action=action)
+            if len(result['results']) != len(resources):
+                msg = "Failed to {action} alert policies to profile '{description}'," +\
+                      "expected {expected} alert profiles to be {action}ed," +\
+                      "but only {changed} were {action}ed"
+                msg = msg.format(action=action,
+                                 description=profile['description'],
+                                 expected=len(resources),
+                                 changed=result['results'])
+                self.module.fail_json(msg=msg)
+        except Exception as e:
+            msg = "Failed to {action} alert policies to profile '{description}': {error}"
+            msg = msg.format(action=action, description=profile['description'], error=e)
+            self.module.fail_json(msg=msg)
+
+        return result['results']
+
+    def update_profile(self, old_profile, new_profile):
+        """ Update alert profile in ManageIQ
+        """
+        changed = False
+        # we need to use client.get to query the alert definitions
+        old_profile = self.client.get(old_profile['href'] + '?expand=alert_definitions')
+
+        # figure out which policies we need to assign / unassign
+        # alert polices listed by the user:
+        new_alert_policies = set(self.get_alert_policies(new_profile['alert_policies']))
+
+        # alert policies which currently exist in the profile
+        if 'alert_definitions' in old_profile:
+            # we use get_policy_href to have a direct href to the policy
+            existing_alert_policies = set([self.get_alert_policy_href(alert) for alert in old_profile['alert_definitions']])
+        else:
+            # no alert policies in this profile
+            existing_alert_policies = set()
+
+        to_add = [dict(href=href) for href in list(new_alert_policies - existing_alert_policies)]
+        to_remove = [dict(href=href) for href in list(existing_alert_policies - new_alert_policies)]
+
+        # assign / unassign the alert policies, if needed
+
+        if to_remove:
```

Comment:
you may accidentally remove policies user did not explicitly specify state: "absent" for policies that currently exist but no the list the user gave.

for example:
if I have policies A and B
and I ask in the playbook to make sure A is present.
the play book may inadvertently delete policy B.

p.s.
you may want to specify new actions to add and remove policies from a profile
e.g. "assign and set policies" ?

cc @cben @joelddiaz ?

Reply:
from my experience with configuration management tools similar to ansible, when you define something using the playbook you *want* it to act like this - to make sure the configuration on the server is 100% in sync with the automated configuration.

"absent" and "present" are the states of the profile itself, when the state is "present" that means "it needs to 1) exist 2) be configured *exactly* how it's described in the playbook".

that's my take on it, but I'll be happy to change it if the ops people decide they want it to behave differently.

Reply:
as someone using this module, i would be surprised if the behavior when creating a new alert profile A, a separate unrelated profile B disappeared.

a different way of saying this is if the product ships with profiles X and Y, then while creating a new profile A, i shouldn't have to provide (or know about) X and Y to the ansible module call when all i want to do is create/edit/delete my own alert profile A.

Reply:
I think you misunderstood the question. This has nothing to do with "unrelated alert profiles", let me paraphrase:

Assume you have manually added an alert profile X, and assigned the alerts A, B and C to it.
If run an ansible playbook that defines alert profile X with only A and B assigned, it will remove C from the existing profile. Is that a kind of behaviour you expect?

If you have a profile Y that is not defined in the playbook, the module will (obviously) not touch it.

Reply:
I absolutely did misunderstand then.

For the primary use case, I think we do want an Ansible call to create/update profile X with alerts A and B to make sure that only alerts A & B are associated with profile X.

There is an edge case where we may want to surgically append/remove alerts from profile X, but I don't think it is critical to support this use case at this time.

Reply:
Okay, so the primary use case is implemented by the module. The edge case of "surgically removing/appending" is not supported at the moment, but should be easy to add in the future if it'll ever be needed.